File: src/ts/reference/core/deployment/existing.ts
import path from "node:path";
import { promises as fs } from "node:fs";
import { parse } from "yaml";
import { pathExists } from "../state/fs";
import type { DeploymentComposeInfo, DeploymentComposePort, DeploymentComposeService } from "./types";

export type ExistingDeploymentFiles = {
  dockerfilePath: string | null;
  composePath: string | null;
};

const composeFileNames = [
  "compose.yaml",
  "compose.yml",
  "docker-compose.yaml",
  "docker-compose.yml",
];

const dockerfileNames = [
  "Dockerfile",
  "dockerfile",
];

export async function findExistingDeploymentFiles(projectRoot: string): Promise<ExistingDeploymentFiles> {
  return {
    dockerfilePath: await findFirstExisting(projectRoot, dockerfileNames),
    composePath: await findFirstExisting(projectRoot, composeFileNames),
  };
}

export async function detectComposePublishedPort(
  composePath: string,
): Promise<{ hostPort: number; containerPort: number } | null> {
  const analysis = await analyzeExistingCompose(composePath);
  const port = selectedComposePort(analysis);
  if (!port?.hostPort) {
    return null;
  }

  return {
    hostPort: port.hostPort,
    containerPort: port.containerPort,
  };
}

export async function analyzeExistingCompose(composePath: string): Promise<DeploymentComposeInfo> {
  const raw = await fs.readFile(composePath, "utf8");
  try {
    const document = parse(raw) as unknown;
    const root = isRecord(document) ? document : {};
    const services = isRecord(root.services) ? root.services : {};
    const serviceEntries = Object.entries(services);

    if (serviceEntries.length === 0) {
      return emptyComposeInfo("Compose file has no services block.");
    }

    const analyzedServices = serviceEntries.map(([name, service]) => analyzeComposeService(name, service));
    const selectedService = selectComposeService(analyzedServices);

    return {
      selectedService: selectedService?.name ?? null,
      serviceReason: selectedService?.reason ?? "No application service could be selected from the Compose file.",
      services: analyzedServices,
      warnings: selectedService?.dependencyLike
        ? [`Selected service ${selectedService.name} looks dependency-like; inspect the existing Compose file before relying on this deployment.`]
        : [],
    };
  } catch (error) {
    return emptyComposeInfo(error instanceof Error ? `Could not parse Compose file: ${error.message}` : "Could not parse Compose file.");
  }
}

export function selectedComposePort(info: DeploymentComposeInfo): DeploymentComposePort | null {
  const service = info.services.find((candidate) => candidate.name === info.selectedService);
  return service?.ports.find((port) => port.hostPort !== null) ?? service?.ports[0] ?? null;
}

async function findFirstExisting(projectRoot: string, names: string[]): Promise<string | null> {
  for (const name of names) {
    const candidate = path.join(projectRoot, name);
    if (await pathExists(candidate)) {
      return candidate;
    }
  }
  return null;
}

function analyzeComposeService(name: string, service: unknown): DeploymentComposeService {
  const value = isRecord(service) ? service : {};
  const image = typeof value.image === "string" ? value.image : null;
  const ports = parseComposePorts(value.ports);
  const expose = parseExpose(value.expose);
  const dependsOn = parseDependsOn(value.depends_on);
  const profiles = parseStringList(value.profiles);
  const build = value.build !== undefined;
  const dependencyLike = isDependencyLikeService(name, image, ports);
  const signals: string[] = [];
  let score = 0;

  if (isAppServiceName(name)) {
    score += 70;
    signals.push(`service name ${name} looks like an application service`);
  }
  if (build) {
    score += 45;
    signals.push("has build configuration");
  }
  if (ports.some((port) => port.hostPort !== null)) {
    score += 35;
    signals.push("publishes a host port");
  } else if (ports.length > 0) {
    score += 20;
    signals.push("declares service ports");
  }
  if (expose.length > 0) {
    score += 10;
    signals.push("exposes internal ports");
  }
  if (dependsOn.length > 0) {
    score += 5;
    signals.push("depends on other services");
  }
  if (dependencyLike) {
    score -= 90;
    signals.push("looks like an infrastructure dependency");
  }
  if (profiles.some((profile) => /test|ci|debug/i.test(profile))) {
    score -= 20;
    signals.push("is behind a test/debug profile");
  }
  if (signals.length === 0) {
    signals.push("no strong service signals");
  }

  return {
    name,
    score,
    image,
    build,
    ports,
    expose,
    dependsOn,
    profiles,
    dependencyLike,
    reason: signals.join("; "),
  };
}

function selectComposeService(services: DeploymentComposeService[]): DeploymentComposeService | null {
  const sorted = [...services].sort((left, right) => right.score - left.score || left.name.localeCompare(right.name));
  return sorted.find((service) => !service.dependencyLike) ?? sorted[0] ?? null;
}

function parseComposePorts(value: unknown): DeploymentComposePort[] {
  if (!Array.isArray(value)) {
    return [];
  }

  return value
    .map((port) => {
      if (typeof port === "string" || typeof port === "number") {
        return parsePortString(String(port));
      }
      if (isRecord(port)) {
        const target = numericPort(port.target);
        if (!target) {
          return null;
        }
        const published = numericPort(port.published);
        return {
          hostPort: published,
          containerPort: target,
          protocol: typeof port.protocol === "string" ? port.protocol : null,
          raw: JSON.stringify(port),
        };
      }
      return null;
    })
    .filter((port): port is DeploymentComposePort => Boolean(port));
}

function parsePortString(raw: string): DeploymentComposePort | null {
  const trimmed = raw.trim();
  const protocolMatch = trimmed.match(/\/([a-z]+)$/i);
  const protocol = protocolMatch?.[1] ?? null;
  const withoutProtocol = trimmed.replace(/\/[a-z]+$/i, "");
  const parts = withoutProtocol.split(":");
  const numericParts = parts
    .map((part) => numericPort(part))
    .filter((port): port is number => port !== null);

  if (numericParts.length === 0) {
    return null;
  }

  const containerPort = numericParts[numericParts.length - 1];
  const hostPort = numericParts.length >= 2 ? numericParts[numericParts.length - 2] : null;
  return {
    hostPort,
    containerPort,
    protocol,
    raw: trimmed,
  };
}

function parseExpose(value: unknown): number[] {
  return parseStringList(value)
    .map((entry) => numericPort(entry))
    .filter((port): port is number => port !== null);
}

function parseDependsOn(value: unknown): string[] {
  if (Array.isArray(value)) {
    return value.filter((entry): entry is string => typeof entry === "string");
  }
  if (isRecord(value)) {
    return Object.keys(value);
  }
  return [];
}

function parseStringList(value: unknown): string[] {
  if (Array.isArray(value)) {
    return value.map(String);
  }
  if (typeof value === "string" || typeof value === "number") {
    return [String(value)];
  }
  return [];
}

function numericPort(value: unknown): number | null {
  if (typeof value === "number" && Number.isInteger(value) && value > 0) {
    return value;
  }
  if (typeof value !== "string") {
    return null;
  }
  const match = value.match(/\d{2,5}/);
  if (!match) {
    return null;
  }
  const port = Number(match[0]);
  return Number.isInteger(port) && port > 0 && port <= 65_535 ? port : null;
}

function isAppServiceName(name: string): boolean {
  return /^(app|web|api|server|backend|frontend|client|www|site|gateway|service)$/i.test(name);
}

function isDependencyLikeService(
  name: string,
  image: string | null,
  ports: DeploymentComposePort[],
): boolean {
  const text = `${name} ${image ?? ""}`.toLowerCase();
  if (/(postgres|postgresql|mysql|mariadb|redis|mongo|rabbitmq|elasticsearch|opensearch|minio|memcached|kafka|zookeeper|localstack|mailhog|mailpit|meilisearch|typesense|clickhouse|influxdb|neo4j|mssql|sqlserver)/.test(text)) {
    return true;
  }
  return ports.some((port) => [5432, 3306, 6379, 27017, 5672, 9200, 9000, 11211, 9092, 2181].includes(port.containerPort));
}

function emptyComposeInfo(reason: string): DeploymentComposeInfo {
  return {
    selectedService: null,
    serviceReason: reason,
    services: [],
    warnings: [reason],
  };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: src/rust/deploy/existing.rs
use std::{fs, path::Path};

use contracts::{DeploymentComposeInfo, DeploymentComposePort, DeploymentComposeService};
use serde_yaml::Value;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ExistingDeploymentFiles {
    pub dockerfile_path: Option<std::path::PathBuf>,
    pub compose_path: Option<std::path::PathBuf>,
}

const COMPOSE_FILE_NAMES: &[&str] = &[
    "compose.yaml",
    "compose.yml",
    "docker-compose.yaml",
    "docker-compose.yml",
];

const DOCKERFILE_NAMES: &[&str] = &["Dockerfile", "dockerfile"];

pub fn find_existing_deployment_files(root: &Path) -> ExistingDeploymentFiles {
    ExistingDeploymentFiles {
        dockerfile_path: find_first_existing(root, DOCKERFILE_NAMES),
        compose_path: find_first_existing(root, COMPOSE_FILE_NAMES),
    }
}

pub fn analyze_existing_compose(compose_path: &Path) -> DeploymentComposeInfo {
    let raw = match fs::read_to_string(compose_path) {
        Ok(raw) => raw,
        Err(error) => return empty_compose_info(format!("Could not read Compose file: {error}")),
    };
    let document = match serde_yaml::from_str::<Value>(&raw) {
        Ok(document) => document,
        Err(error) => return empty_compose_info(format!("Could not parse Compose file: {error}")),
    };
    let Some(services) = document.get("services").and_then(Value::as_mapping) else {
        return empty_compose_info("Compose file has no services block.".to_string());
    };
    let mut analyzed = Vec::new();
    for (name, service) in services {
        let Some(name) = name.as_str() else {
            continue;
        };
        analyzed.push(analyze_compose_service(name, service));
    }
    if analyzed.is_empty() {
        return empty_compose_info("Compose file has no named services.".to_string());
    }
    let selected = select_compose_service(&analyzed);
    DeploymentComposeInfo {
        selected_service: selected.map(|service| service.name.clone()),
        service_reason: selected
            .map(|service| service.reason.clone())
            .unwrap_or_else(|| {
                "No application service could be selected from Compose.".to_string()
            }),
        warnings: selected
            .filter(|service| service.dependency_like)
            .map(|service| {
                vec![format!(
                    "Selected service {} looks dependency-like; generated fallback may be safer.",
                    service.name
                )]
            })
            .unwrap_or_default(),
        services: analyzed,
    }
}

pub fn selected_compose_port(info: &DeploymentComposeInfo) -> Option<DeploymentComposePort> {
    let selected = info.selected_service.as_ref()?;
    let service = info
        .services
        .iter()
        .find(|service| &service.name == selected)?;
    service
        .ports
        .iter()
        .find(|port| port.host_port.is_some())
        .cloned()
        .or_else(|| service.ports.first().cloned())
}

fn find_first_existing(root: &Path, names: &[&str]) -> Option<std::path::PathBuf> {
    names
        .iter()
        .map(|name| root.join(name))
        .find(|path| path.exists())
}

fn empty_compose_info(reason: String) -> DeploymentComposeInfo {
    DeploymentComposeInfo {
        selected_service: None,
        service_reason: reason,
        services: vec![],
        warnings: vec![],
    }
}

fn analyze_compose_service(name: &str, service: &Value) -> DeploymentComposeService {
    let image = service
        .get("image")
        .and_then(Value::as_str)
        .map(str::to_string);
    let build = service.get("build").is_some();
    let ports = parse_ports(service.get("ports"));
    let expose = parse_expose(service.get("expose"));
    let depends_on = parse_depends_on(service.get("depends_on"));
    let profiles = parse_string_list(service.get("profiles"));
    let dependency_like = is_dependency_like_service(name, image.as_deref(), &ports);
    let mut signals = Vec::new();
    let mut score = 0;

    if is_app_service_name(name) {
        score += 70;
        signals.push(format!(
            "service name {name} looks like an application service"
        ));
    }
    if build {
        score += 45;
        signals.push("has build configuration".to_string());
    }
    if ports.iter().any(|port| port.host_port.is_some()) {
        score += 35;
        signals.push("publishes a host port".to_string());
    } else if !ports.is_empty() {
        score += 20;
        signals.push("declares service ports".to_string());
    }
    if !expose.is_empty() {
        score += 10;
        signals.push("exposes internal ports".to_string());
    }
    if !depends_on.is_empty() {
        score += 5;
        signals.push("depends on other services".to_string());
    }
    if dependency_like {
        score -= 90;
        signals.push("looks like an infrastructure dependency".to_string());
    }
    if profiles
        .iter()
        .any(|profile| contains_any(profile, &["test", "ci", "debug"]))
    {
        score -= 20;
        signals.push("is behind a test/debug profile".to_string());
    }
    if signals.is_empty() {
        signals.push("no strong service signals".to_string());
    }

    DeploymentComposeService {
        name: name.to_string(),
        score,
        image,
        build,
        ports,
        expose,
        depends_on,
        profiles,
        dependency_like,
        reason: signals.join("; "),
    }
}

fn select_compose_service(
    services: &[DeploymentComposeService],
) -> Option<&DeploymentComposeService> {
    let mut sorted = services.iter().collect::<Vec<_>>();
    sorted.sort_by(|left, right| {
        right
            .score
            .cmp(&left.score)
            .then_with(|| left.name.cmp(&right.name))
    });
    sorted
        .iter()
        .copied()
        .find(|service| !service.dependency_like)
        .or_else(|| sorted.first().copied())
}

fn parse_ports(value: Option<&Value>) -> Vec<DeploymentComposePort> {
    let Some(Value::Sequence(items)) = value else {
        return vec![];
    };
    items.iter().filter_map(parse_port).collect()
}

fn parse_port(value: &Value) -> Option<DeploymentComposePort> {
    match value {
        Value::String(raw) => parse_port_string(raw),
        Value::Number(number) => parse_port_string(&number.to_string()),
        Value::Mapping(mapping) => {
            let target = mapping
                .get(Value::String("target".to_string()))
                .and_then(numeric_port)?;
            let published = mapping
                .get(Value::String("published".to_string()))
                .and_then(numeric_port);
            let protocol = mapping
                .get(Value::String("protocol".to_string()))
                .and_then(Value::as_str)
                .map(str::to_string);
            Some(DeploymentComposePort {
                host_port: published,
                container_port: target,
                protocol,
                raw: serde_yaml::to_string(value).unwrap_or_else(|_| "port".to_string()),
            })
        }
        _ => None,
    }
}

fn parse_port_string(raw: &str) -> Option<DeploymentComposePort> {
    let trimmed = raw.trim();
    let (without_protocol, protocol) = trimmed
        .rsplit_once('/')
        .map(|(port, protocol)| (port, Some(protocol.to_string())))
        .unwrap_or((trimmed, None));
    let numeric_parts = without_protocol
        .split(':')
        .filter_map(|part| part.parse::<u16>().ok())
        .collect::<Vec<_>>();
    let container_port = *numeric_parts.last()?;
    let host_port = (numeric_parts.len() >= 2).then(|| numeric_parts[numeric_parts.len() - 2]);
    Some(DeploymentComposePort {
        host_port,
        container_port,
        protocol,
        raw: trimmed.to_string(),
    })
}

fn parse_expose(value: Option<&Value>) -> Vec<u16> {
    parse_string_list(value)
        .iter()
        .filter_map(|item| item.parse::<u16>().ok())
        .collect()
}

fn parse_depends_on(value: Option<&Value>) -> Vec<String> {
    match value {
        Some(Value::Sequence(items)) => items
            .iter()
            .filter_map(Value::as_str)
            .map(str::to_string)
            .collect(),
        Some(Value::Mapping(mapping)) => mapping
            .keys()
            .filter_map(Value::as_str)
            .map(str::to_string)
            .collect(),
        _ => vec![],
    }
}

fn parse_string_list(value: Option<&Value>) -> Vec<String> {
    match value {
        Some(Value::Sequence(items)) => items
            .iter()
            .filter_map(|item| {
                item.as_str()
                    .map(str::to_string)
                    .or_else(|| item.as_i64().map(|value| value.to_string()))
            })
            .collect(),
        Some(Value::String(item)) => vec![item.to_string()],
        Some(Value::Number(number)) => vec![number.to_string()],
        _ => vec![],
    }
}

fn numeric_port(value: &Value) -> Option<u16> {
    value
        .as_u64()
        .and_then(|value| u16::try_from(value).ok())
        .or_else(|| value.as_str()?.parse::<u16>().ok())
}

fn is_app_service_name(name: &str) -> bool {
    contains_any(
        name,
        &[
            "app", "web", "frontend", "backend", "api", "server", "service",
        ],
    )
}

fn is_dependency_like_service(
    name: &str,
    image: Option<&str>,
    ports: &[DeploymentComposePort],
) -> bool {
    let combined = format!("{} {}", name, image.unwrap_or_default()).to_ascii_lowercase();
    contains_any(
        &combined,
        &[
            "postgres",
            "mysql",
            "redis",
            "mongo",
            "rabbit",
            "elasticsearch",
            "minio",
            "kafka",
        ],
    ) && !ports.iter().any(|port| port.host_port.is_some())
}

fn contains_any(value: &str, needles: &[&str]) -> bool {
    let lower = value.to_ascii_lowercase();
    needles.iter().any(|needle| lower.contains(needle))
}
